lib/videojs-hlsjs-plugin.js

```diff
@@ -81,40 +82,94 @@ var registerSourceHandler = function (videojs) {
+                    qualitySwitchCallback: switchQuality
                 };
-                cleanTracklist.push(autoLevel);
+
+                tech.trigger('loadedqualitydata', payload);
+
+                // Self-de-register so we don't raise the payload multiple times
+                _video.removeEventListener('playing', _notifyVideoQualities);
             }
+        }
 
-            data.levels.forEach(function (level, index) {
-                var quality = {}; // Don't write in level (shared reference with Hls.js)
-                quality.id = index;
-                quality.selected = index === _hls.manualLevel;
-                quality.label = _levelLabel(level);
+        function _updateHlsjsAudioTrack() {
+            var playerAudioTracks = tech.audioTracks();
+            for (var j = 0; j < playerAudioTracks.length; j++) {
+                if (playerAudioTracks[j].enabled) {
+                    _hls.audioTrack = j;
+                    break;
+                }
+            }
+        }
 
-                cleanTracklist.push(quality);
-            });
+        function _onAudioTracks() {
+            var hlsAudioTracks = _hls.audioTracks;
+            var playerAudioTracks = tech.audioTracks();
+            if (hlsAudioTracks.length > 1 && playerAudioTracks.length === 0) {
+                // Add Hls.js audio tracks if not added yet
+                for (var i = 0; i < hlsAudioTracks.length; i++) {
+                    playerAudioTracks.addTrack(new videojs.AudioTrack({
+                        id: i,
+                        kind: 'alternative',
+                        label: hlsAudioTracks[i].name || hlsAudioTracks[i].lang,
+                        language: hlsAudioTracks[i].lang,
+                        enabled: i === _hls.audioTrack
+                    }));
+                }
 
-            var payload = {
-                qualityData: { video: cleanTracklist },
-                qualitySwitchCallback: switchQuality
-            };
+                // Handle audio track change event
+                playerAudioTracks.addEventListener('change', _updateHlsjsAudioTrack);
+            }
+        }
+
+        function _updateHlsjsTextTrack() {
+            var playerTextTracks = _player.textTracks();
+            var activeTrack = null;
+            for (var j = 0; j < playerTextTracks.length; j++) {
+                if (playerTextTracks[j].mode === 'showing') {
+                    activeTrack = playerTextTracks[j];
+                    break;
+                }
+            }
 
-            tech.trigger('loadedqualitydata', payload);
+            var hlsjsTracks = _video.textTracks;
+            for (var k = 0; k < hlsjsTracks.length; k++) {
+                if (hlsjsTracks[k].kind === 'subtitles' || hlsjsTracks[k].kind === 'captions') {
+                    hlsjsTracks[k].mode = activeTrack !== null
+                        && hlsjsTracks[k].language === activeTrack.language ? 'showing' : 'hidden';
```

Comment:
What if several tracks have the same langage?
Hard to read the way you split it on multiple lines. Maybe add parenthesis arounf the condition so it's clearer?

Reply:
Well JW and Shaka relies on language so it is pretty much assumed to be different.